lib/trie/proof.go

```diff
@@ -28,117 +26,38 @@ import (
 var (
 	// ErrEmptyTrieRoot occurs when trying to craft a prove with an empty trie root
 	ErrEmptyTrieRoot = errors.New("provided trie must have a root")
-
-	// ErrEmptyNibbles occurs when trying to prove or valid a proof to an empty key
-	ErrEmptyNibbles = errors.New("empty nibbles provided from key")
 )
 
-// GenerateProof constructs the merkle-proof for key. The result contains all encoded nodes
-// on the path to the key. Returns the amount of nodes of the path and error if could not found the key
-func (t *Trie) GenerateProof(key []byte, db chaindb.Writer) (int, error) {
-	key = keyToNibbles(key)
-	if len(key) == 0 {
-		return 0, ErrEmptyNibbles
-	}
-
-	var nodes []node
-	currNode := t.root
-
-proveLoop:
-	for {
-		switch n := currNode.(type) {
-		case nil:
-			return 0, errors.New("no more paths to follow")
-
-		case *leaf:
-			nodes = append(nodes, n)
-
-			if bytes.Equal(n.key, key) {
-				break proveLoop
-			}
-
-			return 0, errors.New("leaf node doest not match the key")
+// GenerateProof receive the keys to proof, the trie root and a reference to database
+// will
+func GenerateProof(root []byte, keys [][]byte, db chaindb.Database) ([][]byte, error) {
+	trackedProofs := make(map[string][]byte)
```

Comment:
does this work with multiple keys at the moment?

Reply:
yep!

Reply:
what if multiple node proofs overlap? does this return the same as proof substrate? I'm just trying to understand how this works for multiple proofs within one map (since there's no ordering)